### PR TITLE
sql: make UDF more robust with db/schema renaming

### PR DESCRIPTION
**Pull request by chengxiong-ruan**
There are 5 commits:
1. disallow cross-db references in udf. Also a side effect to serialize user defined type names in udf body as type OIDs.
2. check udf dependencies before renaming a db.
3. check dependencies before renaming a schema.
4. deserialize sequence OID for `SHOW CREATE FUNCTION`.
5. deserialize type OID for `SHOW CREATE FUNCTION`.

Release justification: necessary and low risk bug fixes to make UDF more robust with schema/db renaming.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/udf [hunk 1]
```diff
@@ -1211,6 +1211,66 @@ CREATE FUNCTION public.f_test_cor(IN a INT8, IN b INT8)
...
+SELECT @2 FROM [SHOW CREATE FUNCTION f_udt_rewrite];
+----
+CREATE FUNCTION public.f_udt_rewrite()
+    RETURNS test.public.notmyworkday
+    VOLATILE
+    NOT LEAKPROOF
+    CALLED ON NULL INPUT
+    LANGUAGE SQL
+    AS $$
+    SELECT b'@':::@100115;
```
Comment (chengxiong-ruan) [inline pkg/sql/logictest/testdata/logic_test/udf:1229]: oops.. this is bad.. need to reverse the rewrite when doing `show create`...
will fix this

**Review by ajwerner — COMMENTED** [review]
It seems to me like the rewrite is at least as important. Can you add a bunch of testing related to renaming the schema, database, and type name of a referenced type and make sure that the udf does not break? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-:-NAplNKo3N9-lQe9VLyG:b8gg4pk)* status: :shipit: complete! 0 of 0 LGTMs obtained

File: pkg/sql/logictest/testdata/logic_test/udf (hunk 1 above)
Comment (chengxiong-ruan) [reply pkg/sql/logictest/testdata/logic_test/udf]: this is fixed partially in the last commit. also created #87475 to track the issue.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, 6 of 6 files at r3, 5 of 5 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-:-NBNK3DiEWGXhqmtRb2V:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/rename_database.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNIruBASFNtldRWgVq:-NBNIruCE8TEE_A-WGgi:bc3xy7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/rename_database.go#L158)):*
> ```Go
> 			// Since we now only reference sequences with IDs, it's always safe to
> 			// rename a db or schema containing the sequence.
> 			if tbDesc.IsSequence() {
> ```

you're going to hate me -- this isn't quite true. You can use a sequence like a table, and it won't be rewritten. We may need to file yet another issue for that and we may want to find a way to disallow it. 



_Code quote:_
```Go
			// Since we now only reference sequences with IDs, it's always safe to
			// rename a db or schema containing the sequence.
			if tbDesc.IsSequence() {
```
___
*[`pkg/sql/schema_resolver.go` line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNJ6Y3AacOH2puHJ6j:-NBNJ6Y3AacOH2puHJ6k:bo2r7g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/schema_resolver.go#L241)):*
> ```Go
> 	}
> 	scDesc, err := sr.descCollection.GetImmutableSchemaByID(ctx, sr.txn, fnDesc.GetParentSchemaID(),
> 		tree.SchemaLookupFlags{
> ```

why is this one not `Required`? In practice, it does not matter, but it still seems weird to me. My preference would be that you define a flags variable and cast it

**Review by chengxiong-ruan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87203)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[`pkg/sql/rename_database.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNIruBASFNtldRWgVq:-NBO1yzB1BWbvUdDKse5:b-h66qfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/rename_database.go#L158)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

you're going to hate me -- this isn't quite true. You can use a sequence like a table, and it won't be rewritten. We may need to file yet another issue for that and we may want to find a way to disallow it. 
</blockquote></details>

leaving it as is and filed #87509 to track this. note that this is also broken on master.
Also added test cases for this bad behavior. Thanks.
___
*[`pkg/sql/schema_resolver.go` line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNJ6Y3AacOH2puHJ6j:-NBO2Mol1ECAiydXKFrL:b-j94wx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/schema_resolver.go#L241)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why is this one not `Required`? In practice, it does not matter, but it still seems weird to me. My preference would be that you define a flags variable and cast it
</blockquote></details>

Good point, fixed, thanks!

**Comment by chengxiong-ruan** [comment]
TFTR!
bors r+

**Comment by ajwerner** [comment]
blathers backport 22.2

**Pull request merged**